### ISSUE TITLE
Is there any randomness in sparse quantization?

### ISSUE THREAD

**Issue by somebodyus** [NVIDIA/MinkowskiEngine#188]
I find that my testing results vary from each other. Is there any randomness in the sparse quantization? or any randomness in the convolutional, batch normalization, or relu functions?

**Comment by chrischoy** [comment]
Theres no randomness. But I had a similar experience when I did not set eval mode on a network during the test time since batch norm updates its running mean and var by default.

**Comment by somebodyus** [comment]
I note that a module used contains `ME.MinkowskiDropout(0.5)`. Will it introduce randomness during testing?

**Comment by chrischoy** [comment]
Please put a reproducible code here.

**Comment by somebodyus** [comment]
Thank you for your reply. 
I found that setting the same batch size during testing can give the exactly same results. This confirms that there is no randomness. 
Different batch size will give different results. Moreover, the inference speed per frame also varies a lot when I set different batch sizes. I am wondering is there any influence from the batch size to the sparse-tensor functions, such as batch normalization?

**Comment by somebodyus** [comment]
My tentative explanation to "different batch size gives different testing results": 

Different from a batch of PyTorch data, a batch of sparse tensor data is actually a large data with batch size 1. So it actually modifies the input data, which leads to different testing results.
For example, for a batch of PyTorch data `B*N*3 `(`B` is the batch size, `N` is the number of point per frame, `3` is three coordinates), the network is actually batch-wisely process `B` times of `N*3` data. But for sparse tensor, it rebuilds the data into `1*BN*3` (`1` is the batch size, `BN` is the multiplication of `B` and `N`), so the input is actually modified. If `B` is different, the input is different, and hence the testing results are surely different.

I don't know whether it is correct. @chrischoy

**Comment by chrischoy** [comment]
Your assumption in incorrect. Both pytorch and MinkowskiEngine process data jointly. Pytorch does not process N * 3 data B times.

Regarding the speed depending on the batch size. Of course it varies since you can saturate all CUDA cores with large batch size. With smaller batch sizes, you are using a fraction of CUDA cores and the rest of the CUDA cores are wasted.

Since there's no reproducible code, this issue will be closed, but please feel free to submit a new issue if you have a minimally reproducible code.
